distribution/lib/Standard/Base/0.2.32-SNAPSHOT/src/Data/Vector.enso

```diff
@@ -961,3 +960,17 @@ Singleton_Error.to_display_text : Text
 Singleton_Error.to_display_text =
     "The vector " + this.vec.to_text  + " has only one element."
 
+type Slice
+    type Slice backing offset length
+
+    at : Integer -> Any ! Index_Out_Of_Bounds_Error
+    at index =
+        if ((index < 0) || (index > this.length)) then Error.throw (Index_Out_Of_Bounds_Error index this.length) else
```

Comment:
We have already shortly discussed this, but I'm still not 100% sure if this check should be here. This makes us do the check twice in regular `at` and adds a check to `unsafe_at`.

I know that the idea is to disallow 'looking outside' of the Slice and that is definitely a good idea. But maybe we should just make `Slice` PRIVATE (since it is in fact an implementation detail, so it shouldn't be exposed as part of our interface) and then not care about external user interface. In our implementation we will guarantee that we are only looking inside - there is already a check in `at` and `unsafe_at` is also guaranteed to be called only with valid indices.

Although I do appreciate that with the array an OOB access would raise a panic and with the Slice it can return unexpected values - so given that we are very careful about data corruption bugs, maybe after all it is safer to do the "double-check".

I'm also thinking if this maybe should be a Panic instead of a dataflow error? Then it'd be more consistent with how native Arrays work. After all it is ran without a check only in `unsafe_at` (where now we'll have inconsistent behavior on invalid data - one will throw a panic, another will raise dataflow error - that is maybe ok since this is an unsafe function and the error should never happen anyway) and in `at` there is a higher-level check which will yield dataflow errors anyway.

Reply:
If you want to retain the semantics of this being a slice, you can't really avoid it.